File: runtime/go/units/wasi/fd_renumber.go
// Both endpoints must be live descriptors, else BADF (matching the reference
// table.renumber). The descriptor being overwritten is released, except stdio,
// which must never be closed, and a preopen/dir, which has no OS handle.
func (w *WASI) wasi_fd_renumber(from, to uint32) uint32 {
    fromEntry, ok := w.fds[from]
    if !ok {
        return wasiBadf
    }
    toEntry, ok := w.fds[to]
    if !ok {
        return wasiBadf
    }
    if f, isFile := toEntry.(*os.File); isFile && !w.isStdio(f) {
        f.Close()
    }
    w.fds[to] = fromEntry
    w.meta[to] = w.meta[from]
    delete(w.fds, from)
    delete(w.meta, from)
    return wasiOk
}

File: runtime/java/units/wasi/fd_renumber.java
// Move the fd `from` onto the number `to`, closing whatever `to` held first
// (ADR-40). Both must currently be open (renumbering onto an invalid target is
// EBADF); the table entry and its rights meta move together, and `from` is
// then closed. Renumbering onto stdio or a preopen is allowed — the target's
// old entry is simply replaced (only a guest-opened file carries a channel to
// close).
int wasi_fd_renumber(int from, int to) {
    if (!fds.containsKey(from) || !fds.containsKey(to)) {
        return WASI_BADF;
    }
    if (from == to) {
        return WASI_OK;
    }
    Object toEntry = fds.get(to);
    if (toEntry instanceof Handle) {
        try {
            ((Handle) toEntry).ch.close();
        } catch (java.io.IOException ex) {
            return WASI_IO;
        }
    }
    fds.put(to, fds.remove(from));
    FdMeta m = meta.remove(from);
    if (m != null) {
        meta.put(to, m);
    } else {
        meta.remove(to);
    }
    return WASI_OK;
}
